### PR TITLE
[ViP-Deeplab] Adds depth training pipeline.

### PR DESCRIPTION
**Pull request by joe-siyuan-qiao**
This PR adds the training pipeline for monocular depth estimation for ViP-DeepLab.

### CODE REVIEW

File: model/decoder/panoptic_deeplab.py [hunk 1]
```diff
@@ -257,7 +267,11 @@ def call(self, features, training=False):
       The dictionary containing the predictions under the specified key.
     """
     x = self.conv_block(features, training=training)
-    return {self._pred_key: self.final_conv(x)}
+    x = self.final_conv(x)
```
Comment (aquariusjay) [inline model/decoder/panoptic_deeplab.py:270]: Discussion for design choice:

I feel we should still return logits here, instead of tf.sigmoid(x), a particular type of activation. We should add the desired activation at the place where you need it, similar to what we did for tf.nn.softmax(semantic_logits).

File: model.proto [hunk 2]
```diff
@@ -63,6 +63,12 @@ message HeadOptions {
   // Set the head convolution type. Support 'depthwise_separable_conv' and
   // 'standard_conv'
   optional string head_conv_type = 3 [default = 'depthwise_separable_conv'];
+  // Set the maximum value of sigmoid. When sigmoid_max > sigmoid_min, the head
```
Comment (aquariusjay) [inline model.proto:66]: Discussion for field names.

Maybe a better name for these fields is something like
{min,max}_value_after_activation
because theoretically, they can be used for any type of activation (not just sigmoid).
(any proposal for other name?)
What do you think?

Additionally, please update the comment (saying that these two fields are used to shift the activation values) and probably also mention one example (which is the depth prediction you are adding).

File: model/decoder/vip_deeplab_decoder.py [hunk 3]
```diff
@@ -81,6 +82,18 @@ def __init__(self,
...
+      self._depth_head = panoptic_deeplab.PanopticDeepLabSingleHead(
+          vip_deeplab_options.depth_head.head_channels,
+          vip_deeplab_options.depth_head.output_channels,
+          common.PRED_DEPTH_KEY,
+          name='depth_head',
+          conv_type=vip_deeplab_options.depth_head.head_conv_type,
+          bn_layer=bn_layer,
+          sigmoid_max=vip_deeplab_options.depth_head.sigmoid_max,
+          sigmoid_min=vip_deeplab_options.depth_head.sigmoid_min)
+
```
Comment (aquariusjay) [inline model/decoder/vip_deeplab_decoder.py:96]: We also need to update the function, def checkpoint_items(self), below.

File: model/loss/base_loss.py [hunk 4]
```diff
@@ -557,3 +557,57 @@ def call(self, y_true: Dict[Text, tf.Tensor],
         weights)
     # Reduce_sum over the channels (i.e., number of masks).
     return tf.reduce_sum(weighted_dice_losses, axis=-1)
+
+
+class DepthLoss(tf.keras.losses.Loss):
+  """This class contains code to compute the depth loss.
+
+  This depth loss function combines the scale invariant logarithmic error and
```
Comment (aquariusjay) [inline model/loss/base_loss.py:565]: We should rename the class to better reflect that (it combines the scale invariant logarithmic error and relative squared error). Also, should we have two classes (one for each)?

Please add reference paper for those losses.

**Review by aquariusjay — COMMENTED** [review]
Thanks, Siyuan!

A few things for discussion. PTAL.

**Comment by joe-siyuan-qiao** [comment]
@aquariusjay Thanks for the suggestions! The codes have been revised accordingly.

File: data/preprocessing/input_preprocessing.py [hunk 5]
```diff
@@ -233,18 +246,32 @@ def preprocess_image_and_label(image,
       processed_prev_image.set_shape([crop_height, crop_width, 3])
       if prev_label is not None:
         prev_label.set_shape([crop_height, crop_width, 1])
+    if depth is not None:
+      # Unlabeled depth is 0. We use 0 as the ignore label.
+      _, depth = _pad_image_and_label(
+          image_before_padding, depth, offset_height, offset_width, crop_height,
+          crop_width, 0)
```
Comment (aquariusjay) [inline data/preprocessing/input_preprocessing.py:253]: nit: Use a global variable for depth ignore value 0.

File: data/sample_generator.py [hunk 6]
```diff
@@ -149,6 +149,7 @@ def __call__(self, sample_dict):
       - `prev_label`: An optional tensor of the same shape as `label`.
       - `next_image`: An optional next-frame tensor of the shape of `image`.
       - `next_label`: An optional next-frame tensor of the shape of `label`.
+      - `depth`: An optional tensor of thesame shape as `label`.
```
Comment (aquariusjay) [inline data/sample_generator.py:152]: - `depth`: An optional tensor of the same shape as `label`.
(one space between thesame)

File: model/loss/base_loss.py [hunk 7]
```diff
@@ -557,3 +557,136 @@ def call(self, y_true: Dict[Text, tf.Tensor],
...
+    Returns:
+      A tensor of shape [batch] containing the loss per sample.
+    """
+    gt = y_true[self._gt_key]
+    pred = y_pred[self._pred_key]
+
+    def _compute_error(loss_input):
+      gt, pred = loss_input
+      # Unlabeled pixels are filled with 0.
+      label_mask = gt > 0
```
Comment (aquariusjay) [inline model/loss/base_loss.py:599]: Maybe share this depth ignore 0 with all the other codes (e.g., define it in common.py)

File: model/loss/base_loss.py [hunk 8]
```diff
@@ -557,3 +557,136 @@ def call(self, y_true: Dict[Text, tf.Tensor],
...
```
Comment (aquariusjay) [inline model/loss/base_loss.py:648]: Same here. Share this depth ignore value 0 with other code.

File: model/loss/base_loss.py [hunk 9]
```diff
@@ -557,3 +557,136 @@ def call(self, y_true: Dict[Text, tf.Tensor],
...
+      return relative_squared_error
+
+    return tf.map_fn(_compute_error, (gt, pred), fn_output_signature=tf.float32)
+
+
+class SILogPlusRelativeSquaredLoss(tf.keras.losses.Loss):
+  """This class contains code to compute depth loss SILog + RelativeSquared.
+
+  This depth loss function combines the scale invariant logarithmic (SILog)
+  error and relative squared error.
```
Comment (aquariusjay) [inline model/loss/base_loss.py:663]: This depth loss function combines the scale invariant logarithmic (SILog) error and relative squared error, which was adopted in the ViP-DeepLab model.

**Review by aquariusjay — COMMENTED** [review]
Looks better!
Thanks, Siyuan.
Only a few more nits.
PTAL.

**Comment by joe-siyuan-qiao** [comment]
@aquariusjay The codes have been revised. Thanks for the comments!

File: data/preprocessing/input_preprocessing.py [hunk 10]
```diff
@@ -26,6 +26,8 @@
 
 _MEAN_PIXEL = [127.5, 127.5, 127.5]
 
+_DEPTH_IGNORE_LABEL = 0
```
Comment (aquariusjay) [inline data/preprocessing/input_preprocessing.py:29]: Add comment for it.

File: model/loss/base_loss.py [hunk 11]
```diff
@@ -20,6 +20,8 @@
 import tensorflow as tf
 from deeplab2.model import utils
 
+_DEPTH_IGNORE_LABEL = 0
```
Comment (aquariusjay) [inline model/loss/base_loss.py:23]: Add comment for this.

Seems that we are having _DEPTH_IGNORE_LABEL in multiple files now.
Could we further unify them at one place?
For example, how about we have a helper function, called get_depth_ignore_label, in utils?

File: model/loss/base_loss.py [hunk 12]
```diff
@@ -557,3 +559,134 @@ def call(self, y_true: Dict[Text, tf.Tensor],
...
+
+    Returns:
+      A tensor of shape [batch] containing the loss per sample.
+    """
+    gt = y_true[self._gt_key]
+    pred = y_pred[self._pred_key]
+
+    def _compute_error(loss_input):
+      gt, pred = loss_input
+      label_mask = gt > _DEPTH_IGNORE_LABEL
```
Comment (aquariusjay) [inline model/loss/base_loss.py:600]: != _DEPTH_IGNORE_LABEL?
(more readable?)
or there could be negative value?

File: model/loss/base_loss.py [hunk 13]
```diff
@@ -557,3 +559,134 @@ def call(self, y_true: Dict[Text, tf.Tensor],
...
```
Comment (aquariusjay) [inline model/loss/base_loss.py:648]: same here

**Review by aquariusjay — COMMENTED** [review]
Thanks, Siyuan!
A few more nits for discussion.
PTAL.

File: model/loss/base_loss.py (hunk 11 above)
Comment (joe-siyuan-qiao) [reply model/loss/base_loss.py:23]: Sure. But these two files do not have a shared import -- they seem quite independent. Which file do you think we should import?

**Comment by joe-siyuan-qiao** [comment]
The codes have been revised. Thanks for the suggestions!

**Review by aquariusjay — COMMENTED** [review]
Looks great on my side.
Thanks, Siyuan!

File: data/preprocessing/input_preprocessing.py [hunk 14]
```diff
@@ -121,13 +121,15 @@ def preprocess_image_and_label(image,
                                crop_width,
                                prev_image=None,
                                prev_label=None,
+                               depth=None,
```
Comment (ClaireXie) [inline data/preprocessing/input_preprocessing.py:124]: Did you also update data/preprocessing/input_preprocessing_test.py?

**Review by ClaireXie — COMMENTED** [review]
Thanks Siyuan! Looks good on my side. Just one comment on test.

**Review by YknZhu — COMMENTED** [review]
code lgtm. this also brings a bit more complexity to the input pipeline tho and might confuse new users quite a bit. any thoughts how could we optimize this?

File: data/preprocessing/input_preprocessing.py (hunk 14 above)
Comment (joe-siyuan-qiao) [reply data/preprocessing/input_preprocessing.py:124]: Thanks for the comment! I have updated the test script.

**Comment by joe-siyuan-qiao** [comment]
@YknZhu That's a good question. It might need code refactoring to support any types of potentially multi-modal data?

**Comment by YknZhu** [comment]
Yes it would be nicer to separate these logics into different classes (rather than many if/else in one place)

**Review by ClaireXie — APPROVED** [review]
LGTM
